plugins/backstage-plugin-coder/src/components/CoderProvider/CoderAuthProvider.tsx

```diff
@@ -136,34 +137,38 @@ function useAuthState(): CoderAuth {
     return () => window.clearTimeout(distrustTimeoutId);
   }, [authState.status]);
 
+  const isAuthenticated = validAuthStatuses.includes(authState.status);
+
   // Sets up subscription to spy on potentially-expired tokens. Can't do this
   // outside React because we let the user connect their own queryClient
   const queryClient = useQueryClient();
   useEffect(() => {
     // Pseudo-mutex; makes sure that if we get a bunch of errors, only one
     // revalidation will be processed at a time
-    let isRevalidatingToken = false;
+    let isRevalidating = false;
 
     const revalidateTokenOnError = async (event: QueryCacheNotifyEvent) => {
       const queryError = event.query.state.error;
+
       const shouldRevalidate =
-        !isRevalidatingToken &&
+        isAuthenticated &&
+        !isRevalidating &&
         BackstageHttpError.isInstance(queryError) &&
         queryError.status === 401;
 
       if (!shouldRevalidate) {
         return;
       }
 
-      isRevalidatingToken = true;
+      isRevalidating = true;
       await queryClient.refetchQueries({ queryKey: sharedAuthQueryKey });
```

Comment:
It's sort of both? The logic spies on any 401s made for the currently-loaded query client, regardless of whether the 401 was for a Coder query or not.

Since we let the user supply their own Query client, the thinking was that even if a 401 triggers on a non-Coder query, it might help to revalidate the token, to handle cases like bad internet connections

Reply:
One of the limitations of storing the auth validity state with React Query is that normally, the current state is tied to one specific API function, with one specific query key

So React Query doesn't automatically know that when a non-auth query function errors out, that the underlying auth is probably invalid, too.

Though now that I think about it, I should be able to update the auth state directly when a Coder query fails. Let me see if I can't get that fixed really quick

Reply:
Think I have this improved now!

Reply:
Actually, never mind. Some of the changes are causing the existing tests to fail.
I'll open a new issue about this, but just to keep the scope of this PR small, I went ahead and reverted the changes

Reply:
> handle cases like bad internet connections

I think they would get some network error rather than a 401.  Getting a 401 should mean the server responded and the token is well and truly invalid, and will never be valid again (at least if the token generating system is done correctly hahaha).

Although, it is interesting to consider the case where maybe they have a proxy in front of Coder, and it is the proxy that returned a 401 and not Coder, and the token is actually still valid.  Buuuuut I guess even in that case the second query will still always fail with a 401, since a 401 is not automatically recoverable.

> regardless of whether the 401 was for a Coder query or not

Ohhhhhhh I see, they could use it to make a request to some other service, and that service might give a 401?  That does sound tricky!  I see the use of the revalidate for this case but I agree it would be cool if it could be scoped to Coder queries.